src/libre/print/ebnf.c

```diff
@@ -0,0 +1,294 @@
+	case AST_EXPR_ALT:
+	case AST_EXPR_MANY:
+		return 0;
+
+	case AST_EXPR_FLAGS:
+		return 0; /* XXX */
+
+	default:
+		assert(0);
+	}
+}
+
+static void
+print_grouped(FILE *f, const struct fsm_options *opt, struct ast_expr *n)
+{
+	assert(f != NULL);
+	assert(opt != NULL);
+	assert(n != NULL);
+
+	if (atomic(n)) {
+		pp_iter(f, opt, n);
+	} else {
+		fprintf(f, "(");
+		pp_iter(f, opt, n);
+		fprintf(f, ")");
+	}
+}
+
+static void
+pp_repeat(FILE *f, const struct fsm_options *opt, struct ast_expr *n,
+	unsigned m)
+{
+	assert(f != NULL);
+	assert(opt != NULL);
+	assert(n != NULL);
+
+	if (m == 0) {
+		return;
+	}
+
+	if (m == 1) {
+		pp_iter(f, opt, n);
+		return;
+	}
+
+	fprintf(f, "%u * ", m);
+	print_grouped(f, opt, n);
+}
+
+static void
+pp_iter(FILE *f, const struct fsm_options *opt, struct ast_expr *n)
+{
+	assert(f != NULL);
+	assert(opt != NULL);
+
+	if (n == NULL) { return; }
+
+	switch (n->t) {
+	case AST_EXPR_EMPTY:
+		break;
+
+	case AST_EXPR_CONCAT:
+		pp_iter(f, opt, n->u.concat.l);
+		fprintf(f, ", ");
+		pp_iter(f, opt, n->u.concat.r);
+		break;
+
+	case AST_EXPR_ALT:
+		pp_iter(f, opt, n->u.alt.l);
+		fprintf(f, " | "); /* XXX: indent */
+		pp_iter(f, opt, n->u.alt.r);
+		break;
+
+	case AST_EXPR_LITERAL:
+		fprintf(f, "\'");
+		ebnf_escputc(f, opt, n->u.literal.c);
+		fprintf(f, "\'");
+		break;
+
+	case AST_EXPR_ANY:
+		fprintf(f, "any");
+		break;
+
+	case AST_EXPR_MANY:
+		fprintf(f, "ANY, { ANY }");
+		break;
+
+	case AST_EXPR_REPEATED: {
+		unsigned delta;
+		unsigned i;
+
+		if (n->u.repeated.high == AST_COUNT_UNBOUNDED) {
+			pp_repeat(f, opt, n->u.repeated.e, n->u.repeated.low);
+			if (n->u.repeated.low > 0) {
+				fprintf(f, ", ");
+			}
+
+			fprintf(f, "{ ");
+			pp_iter(f, opt, n->u.repeated.e);
+			fprintf(f, " }");
+
+			return;
+		}
+
+		if (n->u.repeated.low == 0 && n->u.repeated.high == 1) {
+			fprintf(f, "[ ");
+			pp_iter(f, opt, n->u.repeated.e);
+			fprintf(f, " ]");
+
+			return;
+		}
+
+		pp_repeat(f, opt, n->u.repeated.e, n->u.repeated.low);
+
+		delta = n->u.repeated.high - n->u.repeated.low;
+
+		if (n->u.repeated.low > 0 && delta > 0) {
+			fprintf(f, ", ");
+		}
+
+		for (i = 0; i < delta; i++) {
+			fprintf(f, "[ ");
+			pp_iter(f, opt, n->u.repeated.e);
+
+			if (i + 1 < delta) {
+				fprintf(f, ", ");
+			}
+		}
+		for (i = 0; i < delta; i++) {
+			fprintf(f, " ]");
+		}
+
+		break;
+	}
+
+	case AST_EXPR_CHAR_CLASS:
+		fprintf(f, "(");
+		cc_pp_iter(f, opt, n->u.char_class.cca);
+		fprintf(f, ")");
+		break;
+
+	case AST_EXPR_GROUP:
+		fprintf(f, "(");
+		pp_iter(f, opt, n->u.group.e);
+		fprintf(f, ")");
+		break;
+
+	case AST_EXPR_FLAGS:
+		abort();
+
+	default:
+		assert(0);
+	}
+}
+
+void
+re_ast_print_ebnf(FILE *f, const struct fsm_options *opt,
+	const struct ast_re *ast)
+{
+	assert(f != NULL);
+	assert(opt != NULL);
+	assert(ast != NULL);
+
+	fprintf(f, "e = ");
+
+	pp_iter(f, opt, ast->expr);
+
+	fprintf(f, "\n");
+	fprintf(f, "  ;");
+	fprintf(f, "\n");
+}
+
+static void
```

Comment:
Some of these are duplicated in every output format, would it be worth providing a common default but being able to override them?

Reply:
I'd rather come back and refactor them all together. I didn't want to touch tree.c in particular, while you're nearby.